### ISSUE TITLE
Unhandled Exception When Saving KML

### ISSUE THREAD

**Issue by dtdylanb** [SAHorowitz/MSFS2020-PilotPathRecorder#6]
This tool is great, but it seems like a third of the time I go to save the recorded flight out as a kml file for the Google Earth application, it produces an unhandled exception error. It does the same thing if saving for google earth for web. I thought this behavior might be related to active or full pausing in the game, but the latest flight with no pausing had the same results. See below for the full content of the error message.

```
See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.ArgumentException: '', hexadecimal value 0x10, is an invalid character.
   at System.Xml.XmlUtf8RawTextWriter.InvalidXmlChar(Int32 ch, Byte* pDst, Boolean entitize)
   at System.Xml.XmlUtf8RawTextWriter.WriteElementTextBlock(Char* pSrc, Char* pSrcEnd)
   at System.Xml.XmlUtf8RawTextWriter.WriteString(String text)
   at System.Xml.XmlUtf8RawTextWriterIndent.WriteString(String text)
   at System.Xml.XmlWellFormedWriter.WriteString(String text)
   at SharpKml.Base.Serializer.WriteData(XmlWriter writer, String data)
   at SharpKml.Base.Serializer.WriteElement(XmlWriter writer, XmlNamespaceManager manager, Element element, ElementInfo elementInfo)
   at SharpKml.Base.Serializer.ElementSerializer.SerializeElements(Element element, Type elementType)
   at SharpKml.Base.Serializer.ElementSerializer.SerializeElements(Element element, Type elementType)
   at SharpKml.Base.Serializer.SerializeElements(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.SerializeElement(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.WriteElement(XmlWriter writer, XmlNamespaceManager manager, Element element, ElementInfo elementInfo)
   at SharpKml.Base.Serializer.ElementSerializer.SerializeElements(Element element, Type elementType)
   at SharpKml.Base.Serializer.SerializeElements(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.SerializeElement(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.WriteElement(XmlWriter writer, XmlNamespaceManager manager, Element element, ElementInfo elementInfo)
   at SharpKml.Base.Serializer.ElementSerializer.SerializeElements(Element element, Type elementType)
   at SharpKml.Base.Serializer.SerializeElements(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.SerializeElement(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.WriteElement(XmlWriter writer, XmlNamespaceManager manager, Element element, ElementInfo elementInfo)
   at SharpKml.Base.Serializer.ElementSerializer.SerializeElements(Element element, Type elementType)
   at SharpKml.Base.Serializer.SerializeElements(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.SerializeElement(XmlWriter writer, XmlNamespaceManager manager, Element element)
   at SharpKml.Base.Serializer.Serialize(Element root, Stream stream, XmlWriterSettings settings)
   at SharpKml.Base.Serializer.Serialize(Element root, Stream stream)
   at SharpKml.Engine.KmlFile.Save(Stream stream)
   at FS2020PlanePath.MainPage.CreateKMLButton_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4300.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework/v4.0.30319/mscorlib.dll
----------------------------------------
FS2020PlanePath
    Assembly Version: 1.0.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///F:/Documents/MSFS%20Apps/PilotPathRecorder/FS2020PlanePath.exe
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4270.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4300.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4190.0 built by: NET48REL1LAST_B
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4320.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Data.SQLite
    Assembly Version: 1.0.113.0
    Win32 Version: 1.0.113.0
    CodeBase: file:///F:/Documents/MSFS%20Apps/PilotPathRecorder/System.Data.SQLite.DLL
----------------------------------------
System.Data
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4270.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_32/System.Data/v4.0_4.0.0.0__b77a5c561934e089/System.Data.dll
----------------------------------------
System.Transactions
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_32/System.Transactions/v4.0_4.0.0.0__b77a5c561934e089/System.Transactions.dll
----------------------------------------
System.EnterpriseServices
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_32/System.EnterpriseServices/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.EnterpriseServices.dll
----------------------------------------
Microsoft.FlightSimulator.SimConnect
    Assembly Version: 10.0.61259.0
    Win32 Version: 10.0.61637.0 (FSX-Xpack.20070926-1421)
    CodeBase: file:///F:/Documents/MSFS%20Apps/PilotPathRecorder/Microsoft.FlightSimulator.SimConnect.DLL
----------------------------------------
msvcm80
    Assembly Version: 8.0.50727.9672
    Win32 Version: 8.00.50727.9672
    CodeBase: file:///C:/WINDOWS/WinSxS/x86_microsoft.vc80.crt_1fc8b3b9a1e18e3b_8.0.50727.9672_none_d08f9da24428a513/msvcm80.dll
----------------------------------------
SharpKml.Core
    Assembly Version: 5.1.3.0
    Win32 Version: 5.1.3.0
    CodeBase: file:///F:/Documents/MSFS%20Apps/PilotPathRecorder/SharpKml.Core.DLL
----------------------------------------
Anonymously Hosted DynamicMethods Assembly
    Assembly Version: 0.0.0.0
    Win32 Version: 4.8.4300.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_32/mscorlib/v4.0_4.0.0.0__b77a5c561934e089/mscorlib.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.
```

![error message](https://user-images.githubusercontent.com/79209735/108219730-fee20d00-7103-11eb-9fe3-983ae8d2f30f.jpg)

**Comment by SAHorowitz** [comment]
Hmm. I haven't heard of that before so thanks for posting it.  Can you attach your database to this message?   It is called FlightSimFlights.db.  Then let me know which one you are trying to export and I can track down the error.

**Comment by dtdylanb** [comment]
The flight entry is dated 2/16/2021 10:03:06PM    DCD F-15I Strike Eahle Ra'am

[FlightSimFlightsDB.zip](https://github.com/SAHorowitz/MSFS2020-PilotPathRecorder/files/5998896/FlightSimFlightsDB.zip)

**Comment by SAHorowitz** [comment]
Wanted to give you an update. With the database you provided I am able to reproduce the issue but don't know root cause yet. Hopefully will have fix soon.

**Comment by dtdylanb** [comment]
Thanks for checking it out. I did some flights last night in the same aircraft and all recordings produced the error above. I'm going to try it again tonight in a different aircraft to see if it happens again.

**Comment by SAHorowitz** [comment]
Thanks for your patience.  I believe I figured it out.  The waypoints on some of the flights in your database were full of garbage characters instead of a valid waypoint name. Not sure exactly why -- either MSFS2020 has a bug that gives garbage names sometimes or some other situation I have not encountered before.  Either way, I have added protective code both on the recorder and the KML writer parts of the application.  If you see a waypoint named "Unknown Waypoint" in Google Earth that means the program removed the garbage name and replaced it with that phrase.  Please let me know if it works over next few days and I will publish it as a main update. 

To use this version simply unzip the zip file attached in this message over the existing version on your computer.  When the program starts it will say there is a newer version available, please click no to that message.  Test this by creating KML files with flights you know crashed the program as well as recording new flights.

Thanks.
Steve

[PilotPathRecorder.zip](https://github.com/SAHorowitz/MSFS2020-PilotPathRecorder/files/6026196/PilotPathRecorder.zip)

**Comment by dtdylanb** [comment]
That seems to have it worked! I was able to export out the rest of the flights to kml files, and a flight I did last night with points of interest in the flight plan (unknown waypoints), saved out successfully. A couple of things I noticed recently: Even if you select multiple recordings in the flight list, delete flight from database only deletes one of the selected files. Also, it would be helpful to resize the window to see more of the flights in the list at once.
